# BINOMIAL CRR MODEL EVALUATION 

## CREATED BY : Krishn Patel 

## date : June 12 , 2026

In [1]:
import math

In [3]:
# we are provided rhe following values and greeks for this problem are as followed and to be fixed throughout 

S0 = 100
K = 100
r = 0.06
sigma = 0.25
T = 1  
N = 3  

In [5]:
# 1 time steps 

dt = T/N 
print(dt)

0.3333333333333333


In [9]:
# 2. CRR up and down factor 

u = math.exp(sigma*(math.sqrt(dt)))

print(u)

d = 1/u
print(d)
print(round(u*d, 2))

1.155274025440143
0.8655955020013665
1.0


In [11]:
# 3 risk free probablity p 

p = (math.exp(r*dt) - d)/ (u-d)
print(round(p, 4))

0.5337


In [12]:
# 4 discount factor 

df = math.exp(-r* dt)
print(round(df,4))

0.9802


In [13]:
# THREE STEPS STOCK PRICE TREE 

# S(N,J)= S0*(U**N)*(D**J) 

# TIME 0 
print(S0)

# TIME 1 

s11 = S0*u

s10 = S0*d

# TIME 2 

s22 = S0*u*u

s21 = S0*u*d
s12 = S0*d*u

s20 = S0*d*d 

# TIME 3 

s33 = S0*u*u*u 

s32 = S0*u*u*d

s31 = S0*u*d*d

s30 = S0*d*d*d

print(S0 , s11 , s10 , s22 , s21 , s20 , s33 , s32 , s31 , s30)

100
100 115.52740254401431 86.55955020013664 133.46580738566723 100.00000000000001 74.92555730849976 154.18958055705855 115.52740254401431 86.55955020013666 64.855225391183




### Tree Node Value Reference Table

| Time Step ($t$) | Node Variable | Execution Path | Calculated Stock Price |
| :--- | :--- | :--- | :--- |
| **Step 0** | `S0` | Initial Base Price | **100.00** |
| **Step 1** | `s11` | Up ($u$) | **115.53** |
| | `s10` | Down ($d$) | **86.56** |
| **Step 2** | `s22` | Up-Up ($u^2$) | **133.47** |
| | `s21` | Up-Down ($ud$) | **100.00** |
| | `s20` | Down-Down ($d^2$) | **74.93** |
| **Step 3** | `s33` | Up-Up-Up ($u^3$) | **154.19** |
| | `s32` | Up-Up-Down ($u^2d$) | **115.53** |
| | `s31` | Up-Down-Down ($ud^2$) | **86.56** |
| | `s30` | Down-Down-Down ($d^3$) | **64.86** |



### 3-Step Binomial Stock Price Tree (Top-to-Bottom)

                   [ S0: 100.0000 ]
                          │
           ┌──────────────┴──────────────┐
           ▼                             ▼
   [ s11: 115.5274 ]             [ s10: 86.5596 ]
           │                             │
   ┌───────┴───────┐             ┌───────┴───────┐
   ▼               ▼             ▼               ▼



[ s22: 133.4658 ]       [ s21: 100.0000 ]       [ s20: 74.9256 ]
│                       │                       │
┌───┴───┐               ┌───┴───┐               ┌───┴───┐
▼       ▼               ▼       ▼               ▼       ▼
[ s33 ] [ s32 ]         [ s32 ] [ s31 ]         [ s31 ] [ s30 ]


Final Step 3 Leaf Node Values:

* s33 (Up-Up-Up)     : 154.1896
* s32 (Up-Up-Down)   : 115.5274
* s31 (Up-Down-Down) : 86.5596
* s30 (Down-Down-Down): 64.8552

In [14]:
# 5 CALCULATING PAYOFF A

#PAYOFF AT S33 
ps33 = max(S0 - s33 , 0)

#PAYOFF AT S32 
ps32 = max(S0 - s32 , 0)

#PAYOFF AT S31 
ps31 = max(S0 - s31 , 0)

#PAYOFF AT S30 
ps30 = max(S0 - s30 , 0)

print(ps30 , ps31 , ps32 , ps33)

35.144774608817 13.440449799863345 0 0


In [15]:
# valuation at node 2 

v22 = df* ((p*ps33)+((1-p)*ps32))

v21 = df* ((p*ps32)+((1-p)*ps31))

v20 = df* ((p*ps31)+((1-p)*ps30))

print(v22 , v21 , v20)

0.0 6.142980682706686 23.094310022175765


In [16]:
# valuation on node 1 

v11 = df* ((p*v22)+((1-p)*v21))

v10 = df* ((p*v21)+((1-p)*v20))

print(v11 , v10)

2.807659879693252 13.768975726182491


In [17]:
# valuation at node 0 

v00 = df* ((p*v11)+((1-p)*v10))

print(v00)

7.761952737310375


In [18]:
# call put options parity 

cp = S0 - (K * math.exp(-r*T))

print(cp)

5.823546641575135


In [20]:
# call price payoff 

cp33 = max(s33 - K , 0)

cp32 = max(s32 - K , 0)

cp31 = max(s31 - K , 0)

cp30 = max(s30 - K , 0)

print(cp33 , cp32 , cp31 , cp30)

54.18958055705855 15.527402544014308 0 0


In [21]:
# CALL VALUATION AT NODE 2 

cv22 = df* ((p*cp33)+((1-p)*cp32))

cv21 = df* ((p*cp32)+((1-p)*cp31))

cv20 = df* ((p*cp31)+((1-p)*cp30))

print(cv22 , cv21 , cv20)






35.44594005499169 8.123113352031162 0.0


In [22]:
# call valuation at node 1 

cv11 = df* ((p*cv22)+((1-p)*cv21))

cv10 = df* ((p*cv21)+((1-p)*cv20))

cv00 = df* ((p*cv11)+((1-p)*cv10))


print(cv11 , cv10 , cv00)

22.25611850847523 4.24958201108682 13.5854993788855


In [27]:
# BLACK SCHOLES FORMULA 

d1 = (
    math.log(S0 / K)
    + (r + (sigma**2) / 2) * T
) / (sigma * math.sqrt(T))

d2 = d1 - sigma * math.sqrt(T)

print("d1 =", d1)
print("d2 =", d2)

d1 = 0.365
d2 = 0.11499999999999999


In [30]:
N_d1 = 0.643 
N_d2 = 0.546

call = (S0*N_d1) - (K * (math.exp(-r*T))*N_d2)

print(call)

12.87965646630002


In [31]:
put = call - S0 + (K * math.exp(-r*T))

print(put)

7.056109824724885


In [34]:
# american early excersise 
# at node 3 the european and american having same values 

# at time node 2 

a21 = max(max(K - s21 ,0 ) , v21)

a22 = max(max(K - s22,0 ) , v22)

a20 = max(max(K - s20 ,0 ) , v20)

print(a22 , a21, a20)

0 6.142980682706686 25.074442691500238


In [35]:
# EARLY EXCERSISE AT DOWN DOWN IS OPTIMAL WITH VALUE OF 25.06 RS SINCE INTRENSIC VALUE IS GREATER THAN CONTINUM VALUE  AND IN FURTHER NODES AT TIME 1 AND 0 THE INTRENSIC VALUE IS LESSER THAN CONTINUM VALUE THEREFORE NO EARLY EXCERSISE 

V0 = 8.1567
Vd = 14.5322
Vu = 2.86


In [36]:
# greeks for the american put options 

delta = (Vu - Vd)/ (s11 - s10)
print(delta)

-0.40293632615353037


In [37]:
#delta at up node 

deltau = (v22 - v21)/ (s22-s21)

deltad = (v21-v20)/(s21-s20)

print(deltau, deltad)

-0.18355991271669156 -0.6760401237238765


In [38]:
# gamma 

gamma = (deltau- deltad)/ ((s11-s10)/2)

print(gamma)

0.03400184488383518


In [39]:
# theta 

theta = (((Vu+Vd)/2)- V0) / dt 

print(theta)

1.6181999999999963
